### PR TITLE
Saturate large arguments in double-precision erf

### LINKED ISSUE #108483: tf.math.erf returns nan for inf and -inf, inconsistent with PyTorch/SciPy
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

v2.20.0-rc0-4-g72fbba3d20f 2.20.0

### Custom code

Yes

### OS platform and distribution

_No response_

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

## Description
The TensorFlow function `tf.math.erf(x)` computes the Gauss error function element-wise. However, when the input contains positive or negative infinity (+inf / -inf), the output is nan. This is inconsistent with PyTorch and SciPy, which return 1.0 for +inf and -1.0 for -inf.

## Expected Behavior
For mathematical correctness and consistency with standard libraries, tf.math.erf(+inf) should return 1.0 and tf.math.erf(-inf) should return -1.0.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np
import torch
import math
from scipy import special

# TensorFlow behavior
x_tf = tf.constant([np.inf, -np.inf], dtype=tf.float64)
y_tf = tf.math.erf(x_tf)
print("TensorFlow output:", y_tf)
# tf.Tensor([nan nan], shape=(2,), dtype=float64)

# PyTorch behavior
x_torch = torch.tensor(np.array([np.inf, -np.inf]), dtype=torch.float64)
y_torch = torch.erf(x_torch)
print("PyTorch output:", y_torch)
# tensor([1., -1.], dtype=torch.float64)

# Python math library behavior
print(math.erf(float('inf')))   # 1.0
print(math.erf(float('-inf')))  # -1.0

# SciPy behavior
x_np = np.array([np.inf, -np.inf])
y_scipy = special.erf(x_np)
print("SciPy output:", y_scipy)
# [ 1. -1.]
```

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by vishwakt**
## Summary

Fixes #108483.

`tf.math.erf` returns NaN for `+/-inf` in float64, inconsistent with PyTorch, SciPy and `math.erf`. Investigating on 2.21.0 showed the defect is wider than the report:

- **float64 only.** float16, bfloat16 and float32 already return +/-1 for +/-inf, and `erfc` handles infinity in every dtype.
- **Large finite inputs are affected too.** `erf(1.34e154) = 1.0` but `erf(1.35e154) = NaN`; the breakdown threshold is exactly `sqrt(DBL_MAX) ~= 1.3408e154`, because Eigen's `scalar_erf_op<double>` squares its argument and the intermediate overflows. So an inf-only guard would leave `erf(1e300) = NaN`.

## Fix

Specialize `tensorflow::functor::erf` for double with a saturating wrapper in `cwise_ops.h`: `erf(x)` rounds to exactly +/-1 in double precision for every `|x| >= 6` (verified: the current implementation already returns exactly 1.0 from 5.8 upward), so such arguments are answered directly and everything else forwards to the Eigen implementation unchanged.

Design points, mapped to the feedback on the earlier attempt (#108592, closed by the stale bot while waiting for review):

- **Perf** (@mihaimaruseac's concern about a branch slowing the kernel): the packet path stays fully vectorized, using one compare and select around the existing `packetOp`; the scalar path marks the saturation branch `TF_PREDICT_FALSE`. The float path, which is the hot dtype in practice, is untouched.
- **dtype compatibility** (raised in that PR's review): moot by construction, since only the double instantiation changes; float, half and bfloat16 keep the existing functor.
- **NaN**: the saturation mask is written as `|x| >= 6` selecting the saturated value, so NaN fails the comparison and keeps the forwarded value, preserving `erf(NaN) = NaN` on both scalar and packet paths.
- **No correct value changes**: for `6 <= |x| <= 1.34e154` the current implementation already returns exactly +/-1, so the select replaces 1.0 with 1.0; below 6 nothing changes.

All packet primitives used (`pcmp_le`, `pselect`, `pabs`, `pnegate`, `pzero`) are already used elsewhere in this header.

## Verification

The regression test covers +/-inf, values on both sides of the overflow threshold (1.34e154 / 1.35e154), a large finite value (1e300), the saturation seam (5.9, 6.0, 6.1) and NaN propagation. Against the unpatched 2.21.0 wheel it fails as expected, so it gates the change; the C++ itself is not compiled locally (no Bazel on this machine), so CI is the first build.

Arguably the root fix belongs in Eigen's `erf<double>`; this addresses it at the TF layer as #108592 did, which @mihaimaruseac considered reasonable pending review. Happy to also file the upstream Eigen issue if useful.

Credit to @Thrsu for the report and to @shahid1330 for the earlier attempt in #108592.

### CODE REVIEW

File: tensorflow/python/kernel_tests/math_ops/cwise_ops_unary_test.py [hunk 1]
```diff
@@ -323,6 +323,20 @@ def testFloatEmpty(self):
...
+    # Regression test for GitHub issue 108483. The double-precision erf
+    # squared its argument, so |x| > sqrt(DBL_MAX), about 1.34e154, and
+    # +/-inf overflowed the intermediate and returned NaN. Values on both
+    # sides of that overflow threshold are covered, along with the
+    # saturation seam at 6 and NaN propagation.
+    x = np.array([np.inf, -np.inf, 1e300, -1e300, 1.35e154, -1.35e154,
+                  1.34e154, 5.9, 6.0, 6.1, -7.0], dtype=np.float64)
+    expected = np.array([1.0, -1.0, 1.0, -1.0, 1.0, -1.0,
+                         1.0, 1.0, 1.0, 1.0, -1.0], dtype=np.float64)
+    self.assertAllEqual(expected, self.evaluate(math_ops.erf(x)))
```
[the remark this answers is not in the document]
Comment (vishwakt) [reply tensorflow/python/kernel_tests/math_ops/cwise_ops_unary_test.py]: Split rather than swapped wholesale: the saturating branch returns the literal +/-1.0 by construction, so exact equality cannot flake and those cases keep assertAllEqual, while 5.9 and 1.34e154 exercise the forwarded Eigen implementation whose last-ulp rounding can vary across platforms and now use assertAllClose as suggested.

**Review by dmiltr3 — APPROVED** [review]

**Comment by dmiltr3** [comment]
Hi @vishwakt,

Thank you for the detailed investigation and PR! Your root cause analysis is spot on.

Our numerical ops and Eigen maintainers reviewed this and requested that this fix be implemented directly upstream in Eigen rather than adding a custom functor wrapper in TensorFlow (cwise_ops.h).

**Why:**
In Eigen (unsupported/Eigen/src/SpecialFunctions/SpecialFunctionsImpl.h):

generic_fast_erf<float> clamps x^2 to 16.0f.
generic_fast_erfc<double> clamps x to [-28.0, 28.0].
generic_fast_erf<double> is missing an input clamp before calling erfc_double_large(x, x2), which causes x^2 to overflow to inf and yield NaN for |x| > sqrt(DBL_MAX) and ±inf.

**Next Steps:**
Please open a Merge Request on the upstream Eigen repository at https://gitlab.com/libeigen/eigen to add the appropriate clamping/saturation in generic_fast_erf<double>::run (in unsupported/Eigen/src/SpecialFunctions/SpecialFunctionsImpl.h).
Once accepted upstream in Eigen, TensorFlow will inherit the fix when the Eigen dependency is updated.

Thank you again for diagnosing the exact breakdown threshold and helping improve the numerical stability of both TensorFlow and Eigen!
